pypose/module/pnp.py

```diff
@@ -0,0 +1,489 @@
+        For the details, please refer to equation 15.
+        """
+
+        def __init__(self, betas):
+            super().__init__()
+            self.betas = torch.nn.Parameter(betas)
+
+        def forward(self, ctrl_pts_w, kernel_bases):
+            """
+            Args:
+                ctrl_pts_w: The control points in world coordinate. The shape is (B, 4, 3).
+                kernel_bases: The kernel bases. The shape is (B, 16, 4).
+            Returns:
+                torch.Tensor: The loss. The shape is (B, ).
+            """
+            batch_shape = kernel_bases.shape[:-2]
+            # calculate the control points in camera coordinate
+            ctrl_pts_c = pypose.bmv(kernel_bases, self.betas)
+            diff_c = ctrl_pts_c.reshape(*batch_shape, 1, 4, 3) - ctrl_pts_c.reshape(*batch_shape, 4, 1, 3)
+            diff_c = diff_c.reshape(*batch_shape, 16, 3)
+            diff_c = torch.sum(diff_c ** 2, dim=-1)
+
+            # calculate the distance between control points in world coordinate
+            diff_w = ctrl_pts_w.reshape(*batch_shape, 1, 4, 3) - ctrl_pts_w.reshape(*batch_shape, 4, 1, 3)
+            diff_w = diff_w.reshape(*batch_shape, 16, 3)
+            diff_w = torch.sum(diff_w ** 2, dim=-1)
+
+            error = (diff_w - diff_c).abs().mean(dim=-1)
+
+            return error
+
+    def __init__(self, naive_ctrl_pts=False, optimizer=GaussNewton):
+        super().__init__()
+        self.naive_ctrl_pts = naive_ctrl_pts
+        self.optimizer = optimizer
+
+        self.register_buffer('six_indices', torch.tensor(
+            [(0 * 4 + 1), (0 * 4 + 2), (0 * 4 + 3), (1 * 4 + 2), (1 * 4 + 3), (2 * 4 + 3)]))
+        self.register_buffer('six_indices_pair', torch.tensor([(0, 1), (0, 2), (0, 3), (1, 2), (1, 3), (2, 3)]).T)
+        self.register_buffer('ten_indices_pair', torch.tensor([(0, 0),
+                                                               (0, 1), (1, 1),
+                                                               (0, 2), (1, 2), (2, 2),
+                                                               (0, 3), (1, 3), (2, 3), (3, 3)]).T)
+        # equal mask for above pairs [ True, False, True, False, False, True, False, False, False, True, ]
+        self.register_buffer('multiply_mask', torch.tensor([1., 2., 1., 2., 2., 1., 2., 2., 2., 1.]))
+
+    def forward(self, points, pixels, intrinsics):
+        """
+        Args:
+            points (Tensor): 3D object points in the world coordinates, shape (batch_size, n, 3)
+            pixels (Tensor): 2D image points, which are the projection of object points, with shape (batch_size, n, 2)
+            intrinsics (Tensor): camera intrinsics, shape (batch_size, 3, 3)
+        Returns:
+            lietensor.SE3Type: estimated pose for the camera
+        """
+        # shape checking
+        batch_shape = points.shape[:-2]
+        assert pixels.shape[:-2] == batch_shape
+        assert intrinsics.shape[:-2] == batch_shape
```

Comment:
batch dimension is unnecessary to be the same, they can be broadcast-able, e.g., `(1,2,3)` with `(2,1,1)` is also fine.